src/core/lib/security/credentials/external/pluggable_auth_external_account_credentials.cc

```diff
@@ -0,0 +1,401 @@
+            CreateExecutableResponse(std::string(output_file_content), &error);
+        if (executable_response_ == nullptr) {
+          FinishRetrieveSubjectToken("", error);
+          return;
+        }
+        // If the cached output file has an executable response
+        // that was successful and un-expired, return the subject token.
+        if (executable_response_->success &&
+            !isExpired(executable_response_->expiration_time)) {
+          FinishRetrieveSubjectToken(executable_response_->subject_token,
+                                     absl::OkStatus());
+          return;
+        }
+      }
+    }
+  }
+  // If the cached output_file does not contain a valid response, call the
+  // executable.
+  std::vector<std::string> envp_vector = {
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_AUDIENCE=%s", options.audience),
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_TOKEN_TYPE=%s",
+                      options.subject_token_type),
+      absl::StrFormat("GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE=%d", 0),
+      absl::StrFormat(
+          "GOOGLE_EXTERNAL_ACCOUNT_IMPERSONATED_EMAIL=%s",
+          get_impersonated_email(options.service_account_impersonation_url))};
+  if (output_file_path_ != "")
+    envp_vector.push_back(absl::StrFormat(
+        "GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE=%s", output_file_path_));
+  int environ_count = 0, envc = 0, argc = 0;
+  while (environ[environ_count] != nullptr) environ_count += 1;
+  char** envp =
+      (char**)gpr_malloc(sizeof(char*) * (environ_count + envp_vector.size()));
+  for (; envc < environ_count; envc++) envp[envc] = gpr_strdup(environ[envc]);
+  for (int j = 0; j < envp_vector.size(); envc++, j++)
+    envp[envc] = gpr_strdup(envp_vector[j].c_str());
+  std::vector<std::string> arg_vector = absl::StrSplit(command_, " ");
+  argc = arg_vector.size();
+  char** argv = (char**)malloc(sizeof(char*) * argc);
+  for (int i = 0; i < argc; i++) {
+    argv[i] = gpr_strdup(arg_vector[i].c_str());
+  }
+  std::packaged_task<bool(gpr_subprocess**, int, char**, int, char**,
+                          std::string*, std::string*, std::string*)>
+      run_executable_task(run_executable);
+  std::future<bool> executable_output_future = run_executable_task.get_future();
+  std::string output_string, stderr_data, error_string;
+  std::thread thr(std::move(run_executable_task), &gpr_subprocess_, argc, argv,
+                  envc, envp, &output_string, &stderr_data, &error_string);
+  if (executable_output_future.wait_for(std::chrono::seconds(
+          executable_timeout_ms_ / 1000)) != std::future_status::timeout) {
+    thr.join();
+    if (!executable_output_future.get()) {
+      // An error must have occured in the executable.
+      FinishRetrieveSubjectToken(
+          "", GRPC_ERROR_CREATE(absl::StrFormat(
+                  "Executable failed with error: %s", error_string)));
+      return;
+    }
+    // If the output file is specified, then the executable has stored the
```

Comment:
IIRC even if output_file is specified, they are still expected to write to STDOUT